pkg/innerring/indexer.go

```diff
@@ -0,0 +1,80 @@
+package innerring
+
+import (
+	"crypto/ecdsa"
+	"sync"
+	"time"
+
+	"github.com/nspcc-dev/neofs-node/pkg/innerring/invoke"
+	"github.com/nspcc-dev/neofs-node/pkg/morph/client"
+	"github.com/pkg/errors"
+)
+
+type innerRingIndexer struct {
+	sync.RWMutex
+
+	cli     *client.Client
+	key     *ecdsa.PublicKey
+	timeout time.Duration
+
+	innerRingIndex, innerRingSize int32
+
+	lastAccess time.Time
+}
+
+func newInnerRingIndexer(cli *client.Client, key *ecdsa.PublicKey, to time.Duration) *innerRingIndexer {
+	return &innerRingIndexer{
+		cli:     cli,
+		key:     key,
+		timeout: to,
+	}
+}
+
+func (s *innerRingIndexer) update() (err error) {
+	s.RLock()
+
+	if time.Since(s.lastAccess) < s.timeout {
+		s.RUnlock()
+		return nil
+	}
+
+	s.RUnlock()
+
+	s.Lock()
+	defer s.Unlock()
+
+	if time.Since(s.lastAccess) < s.timeout {
+		return nil
+	}
+
+	s.innerRingIndex, s.innerRingSize, err = invoke.InnerRingIndex(s.cli, s.key)
+	if err != nil {
+		return err
+	}
+
+	s.lastAccess = time.Now()
+
+	return nil
+}
+
+func (s *innerRingIndexer) InnerRingIndex() (int32, error) {
+	if err := s.update(); err != nil {
+		return 0, errors.Wrap(err, "can't update index state")
+	}
+
+	s.RLock()
```

Comment:
As an option, we can return `index` and `size` values from method update
```
func (s *innerRingIndexer) update() (index, size int, err error)
```
 in order to reduce the number of synchronization points. Then the method body can be shortened to
```
func (s *innerRingIndexer) InnerRingIndex() (int32, error) {
	index, _, err := s.update()
	return index, err
}
```
(same for `InnerRingSize` method).

Reply:
Seems reasonable. Instead of ints I decided to return structure that stores all indexes. Check it out: https://github.com/nspcc-dev/neofs-node/pull/446/commits/2052065423ddc76d64feb74bb3f74ad36f632527

Reply:
Good, but do we rly need to return a pointer? It seems like we can just `return s.ind` from `update`.

Reply:
We don't. I was just too lazy to write `return indexes{}, err` instead of `return nil, err`. Fixed.